be/btree.c

```diff
@@ -1105,39 +1233,81 @@ be_btree_get_btree_node(struct m0_be_btree_cursor *it, void *key, bool slant)
 	struct m0_be_btree 	*tree = it->bc_tree;
 	struct m0_be_bnode 	*bnode = tree->bb_root;
 	struct btree_node_pos    bnode_pos = { .bnp_node = NULL };
-
+	
+	char *strkey = (char *)key;
 	it->bc_stack_pos = 0;
+	
+	if(tree->bb_ops->ko_type == M0_BBT_CAS_CTG || tree->bb_ops->ko_type == M0_BBT_CONFDB 
+	||tree->bb_ops->ko_type == M0_BBT_INVALID|| tree->bb_ops->ko_type == M0_BBT_UT_KV_OPS
+	|| tree->bb_ops->ko_type == M0_BBT_NR || tree->bb_ops->ko_type == M0_BBT_COB_FILEATTR_EA
+	|| tree->bb_ops->ko_type == M0_BBT_COB_NAMESPACE||tree->bb_ops->ko_type==M0_BBT_BALLOC_GROUP_EXTENTS
+	|| tree->bb_ops->ko_type == M0_BBT_BALLOC_GROUP_DESC|| tree->bb_ops->ko_type == M0_BBT_EMAP_EM_MAPPING)
+	{
+		while (true) {
+			/*  Retrieve index of the key equal to or greater than */
+			/*  the key being searched */
+			idx = 0;
+			while (idx < bnode->bt_num_active_key &&
+				key_gt(tree, key, bnode->bt_kv_arr[idx].btree_key)) {
+				idx++;
+			}
 
-	while (true) {
-		/*  Retrieve index of the key equal to or greater than */
-		/*  the key being searched */
-		idx = 0;
-		while (idx < bnode->bt_num_active_key &&
-		       key_gt(tree, key, bnode->bt_kv_arr[idx].btree_key)) {
-			idx++;
-		}
+			/*  If key is found, copy key-value pair */
+			if (idx < bnode->bt_num_active_key &&
+				key_eq(tree, key, bnode->bt_kv_arr[idx].btree_key)) {
+				bnode_pos.bnp_node = bnode;
+				bnode_pos.bnp_index = idx;
+				break;
+			}
 
-		/*  If key is found, copy key-value pair */
-		if (idx < bnode->bt_num_active_key &&
-		    key_eq(tree, key, bnode->bt_kv_arr[idx].btree_key)) {
-			bnode_pos.bnp_node = bnode;
-			bnode_pos.bnp_index = idx;
-			break;
+			/*  Return NULL in case of leaf node and did not get key*/
+			if (bnode->bt_isleaf) {
+				while (bnode != NULL && idx == bnode->bt_num_active_key)
+					bnode = node_pop(it, &idx);
+				if (slant && bnode != NULL) {
+					bnode_pos.bnp_node = bnode;
+					bnode_pos.bnp_index = idx;
+				}
+				break;
+			}
+			/*  Move to a child node */
+			node_push(it, bnode, idx);
+			bnode = bnode->bt_child_arr[idx];
 		}
+	}
+	else
+	{
+		while (true) {
+			/*  Retrieve index of the key equal to or greater than */
+			/*  the key being searched */
+			idx = 0;
+			while (idx < bnode->bt_num_active_key &&
+				strcmp(strkey, bnode->bt_kv_arr[idx].inlkey) > 0) {
+				idx++;
+			}
 
-		/*  Return NULL in case of leaf node and did not get key*/
-		if (bnode->bt_isleaf) {
-			while (bnode != NULL && idx == bnode->bt_num_active_key)
-				bnode = node_pop(it, &idx);
-			if (slant && bnode != NULL) {
+			/*  If key is found, copy key-value pair */
+			if (idx < bnode->bt_num_active_key &&
```

Comment:
Is it required to check for _idx < bnode->bt_num_active_key_ ?

Reply:
It is part of the original code.
I havent modified it. 
So it might be necessary to keep that check.